`Communication`: Allow users to communicate in group chats and channels

stefanwaldhauser opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context


`Summary: Extend the communication section of a course to a more fully fledged chat app. This means the user can have this open in a separate browser window and use it similar to other chat apps like Slack or Zulip. It should feel like a standalone application`

This PR improves on the great communication feature implemented by @canberkanar by extending the real-time communication feature of Artemis further:
- Existing: OneToOneChats (implemented by@canberkanar): Between exactly two users
- 🆕 Group Chats: Between up to 10 users
- 🆕 Channels (Private and Public): Unlimited amount of users

### Description

Analysis Object Diagram: 
![image](https://user-images.githubusercontent.com/29718932/208452805-16579ae2-bdc9-4619-98f3-820600a9985a.png)



- The existing **Conversation** entity was renamed to **OneToOneChat** and now represents one child class of the abstract parent class **Conversation**.
- Introduced new conversation child classes **GroupChat** and **Channel**
- Communication between client and server only via DTOs. This allows fine-grained control over what we send compared to sending raw entities, as was done previously.
- The DTOs include what rights the user has for the respective conversation (for example, the right to add or remove users) as transient properties to show and hide UI elements without complicated client logic dynamically. This also makes sure server, and client authorization logic is in sync.
- All UI components dynamically adapt to the concrete type of conversation to minimize code duplication and give a consistent look and feel.
- To preserve "app feel," everything happens via dialogs, and for no action, navigation to a subpage is necessary. 
- Various UI components are kept in sync by using rxjs subjects heavily.
- The client gets notified about conversation updates/deletion via WebSockets so the user does not have to reload the page. For example, he/she will automatically see a new conversation they have been added to in the sidebar. This further preserves the "app feel"


### Steps for Testing
The following features should be tested. It is best to log in with two accounts in separate browsers to test the WebSocket feature.

**I deliberately will not write exactly how to perform each action as I want to know if some of the actions are hard to perform/find, which is very helpful feedback for me for usability purposes **

_All Conversation Types_
- Can hide/unhide conversations in the left sidebar.
- Can favorite/ unfavorite conversations in the left sidebar.
- See an unread badge (message bubble icon) next to the not selected conversation when a new message comes in. This badge should be removed once you open the respective conversation by clicking on it.
- Can search the conversations in the left sidebar by name.
- You can reply in a thread to a message, causing the right-side thread sidebar to open

_OneToOneChats_
- Can be started by any user of a course.
- Up to 2 members.
- Can not be left or deleted.

_GroupChats_
- Can be started by any user of a course.
- Up to 10 members can be added.
- Any member can add/remove users.
- Can be given an explicit name.
- Members can leave a group chat.
- Can not be deleted

_Channel_
- No member limit
- Can be started only by tutors, editors, and instructors
- Can be private (no self-join possible) or public (self-join possible)
- Users can browse for all public channels and self-join them
- Note: Instructors can self-join all channels, public or private
- Users can be promoted to channel moderators (creator automatically is channel moderator)
- Channel moderators can add/remove individual users (except the creator) or whole course groups (like all tutors)
- Channel moderators can delete and edit messages of other users in a channel.
- Channel moderators can grant / revoke channel moderator role (except from the creator)
- Channel archival / unarchival possible for all channel moderators
- Channel deletion is possible only for instructors or the creator
- Channels have name, topic and description that can be changed by channel moderators.
- Members (except the creator) can leave a channel.

Access Control Matrix:
![image](https://user-images.githubusercontent.com/29718932/208550342-537a40f7-827b-4b0d-b590-95f83ff7da31.png)


### Review Progress


#### Code Review
- [ ] Review 1
- [ ] Review 2
- [ ] Review 3
- [ ] Review 4
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
- [ ] Test 3
- [ ] Test 4

### Test Coverage

#### Client (Condensed Folder View as too many individual files)
![image](https://user-images.githubusercontent.com/29718932/205518687-a7eb7cb6-2a4e-4fd6-825b-b0f691288a53.png)


#### Server
|                                                     |    | Line Coverage |    | Branch Coverage |
|-----------------------------------------------------|----|---------------|----|-----------------|
| ChannelResource                                     | ## |           92% | ## |             71% |
| GroupChatResource                                   | ## |           89% | ## |             61% |
| ConversationResource                                | ## |           88% | ## |             50% |
| OneToOneChatResource                                | #  |          100% | #  |            100% |
| ConversationService                                 | ## |           96% | ## |             82% |
| ChannelService                                      | ## |           94% | ## |             66% |
| GroupChatService                                    | ## |           86% | ## |             50% |
| ConversationDTOService                              | ## |           95% | ## |             79% |
| OneToOneChatService                                 | #  |          100% | #  |            100% |
| ConversationService.ConversationMemberSearchFilters | #  |          100% |    |             n/a |
| GroupChatAuthorizationService                       | ## |           78% | ## |             50% |
| ChannelAuthorizationService                         | ## |           98% | ## |             88% |
| OneToOneChatAuthorizationService                    | ## |           84% | ## |             50% |
| ConversationAuthorizationService                    | #  |          100% | ## |             70% |
| ChannelDTO                                          | ## |           81% |    |             n/a |
| OneToOneChatDTO                                     | ## |           72% |    |             n/a |
| ConversationDTO                                     | #  |          100% | ## |             75% |
| GroupChatDTO                                        | #  |          100% |    |             n/a |
| ConversationUserDTO                                 | #  |          100% |    |             n/a |
| GroupChat                                           | ## |           41% | #  |              0% |
| OneToOneChat                                        | ## |           70% | ## |             50% |
| Conversation                                        | #  |          100% |    |             n/a |
| Channel                                             | #  |          100% |    |             n/a |
| ConversationSettings                                | #  |          100% |    |             n/a |
| ConversationRepository                              | ## |           56% |    |             n/a |
| ChannelNameDuplicateException                       | #  |          100% |    |             n/a |
          

### Screenshots

![image](https://user-images.githubusercontent.com/29718932/205522680-79d842d9-4d46-4801-a306-01944906d7e2.png)
![image](https://user-images.githubusercontent.com/29718932/205522710-023593ae-e2c3-4042-b296-f569011f628d.png)
![image](https://user-images.githubusercontent.com/29718932/205522763-10999086-0e14-4062-81b2-7e294d9f0565.png)
![image](https://user-images.githubusercontent.com/29718932/205522776-1095f327-0084-4392-90e8-1d96642672ba.png)
![image](https://user-images.githubusercontent.com/29718932/205522817-83283dc5-4e8d-4ab7-865b-653ad9bd136c.png)
![image](https://user-images.githubusercontent.com/29718932/205524154-dc76ec60-2ad6-4a26-adda-7411018cadaf.png)
![image](https://user-images.githubusercontent.com/29718932/205524171-69d8f5e4-c839-45f2-af6b-3d4750c19f8e.png)
![image](https://user-images.githubusercontent.com/29718932/205524188-6ac05fc8-5c2c-4c14-826f-2770dc679d86.png)
![image](https://user-images.githubusercontent.com/29718932/205524193-ef311dbd-7367-4e53-985f-3e2d72baebc4.png)
![image](https://user-images.githubusercontent.com/29718932/205524210-e53b7df0-c82f-41ea-8f06-f9c9ac3cbedb.png)
![image](https://user-images.githubusercontent.com/29718932/205524232-58e489bb-6c8d-453b-b846-a5c748f821f5.png)
![image](https://user-images.githubusercontent.com/29718932/205524260-f6332378-f387-4f56-acf3-16bb770140d7.png)
![image](https://user-images.githubusercontent.com/29718932/205524292-29449749-ceaa-448a-9f7a-394674a6edc7.png)
![image](https://user-images.githubusercontent.com/29718932/205524364-cd48ffef-9313-4148-b5ed-51ffe26d69db.png)
![image](https://user-images.githubusercontent.com/29718932/205524349-cef87e33-ab7f-4e78-bf35-bdb0f5ecc660.png)
![image](https://user-images.githubusercontent.com/29718932/205524381-e2d57b08-eee5-4a38-868d-4fc60f93dbb8.png)

tobias-lippert reviewed (commented):
Tested in testing session on ts1. Really nice features with great usability.

akesfeden reviewed (commented):
Tested on TS1 in testing session. There are lots of great features, I did not notice any critical functionality missing.

minrows reviewed (commented):
Tested on ts1 during the testing session. Nice feature. 👍 Left minor feedback regarding the appearance of the message indicator to the author.

pellumbbaboci reviewed (commented):
Tested during testing session on TS1. Newly added features are easy to use and there are no critical functionalities missing.

stefanwaldhauser commented:
https://user-images.githubusercontent.com/29718932/205676767-be6d9af5-505c-4a76-8e94-74ba535c78e7.mp4

Small Screencast showing messaging between two users in a channel. Note the new flag for new messages.

ge65cer reviewed (commented):
Wow, nice job! 🎉  All in all it worked pretty well. Here is my general feedback:
* The user interface needs some polishing.
* The difference between channels and groups is a little bit unclear. Are there concrete benefits for the user? Do we need both or are we maybe introducing an additional special case complicating future maintenance?
* Favorites (users/groups/channels) should be shown _both_ under "Favorites" and their respective list.
* I was able to add users to a channel without being the channel admin. (Maybe because I am an Artemis admin?) The separation is unclear, maybe rename to "Channel Owners" or similar.
* Message search: Should work by pressing enter and also highlight the text.
* When pressing enter in the message field, it should be disabled until the request was successful (can still enter text while sending).
* Showing the ID under details is unnecessary for users.
* Wording: Sometimes the UI uses "channel" and other times "conversation"

stefanwaldhauser commented:
> Wow, nice job! 🎉 All in all it worked pretty well. Here is my general feedback:
> 
> * The user interface needs some polishing.
- Can you be more specific about which parts?

> * The difference between channels and groups is a little bit unclear. Are there concrete benefits for the user? Do we need both or are we maybe introducing an additional special case complicating future maintenance?

- It is a common pattern that other chat programs like slack also implement. I see the following reasons why both are valid:

1. Instructors generally want to have control over their course communication. If we allow any user to just open a channel, there is a danger the course will be cluttered by a lot of public channels. 
2. There is no upper limit on the members in a channel. The danger of spam increases if we allow normal users to create channels and invite everybody to the course.
3. They are semantically different: Channels provide a more organized and structured way to communicate around a specific topic such as tech-support. Group chats, on the other hand, are more flexible and can be used for more casual or informal conversations between course users.


> * Favorites (users/groups/channels) should be shown _both_ under "Favorites" and their respective list.

I disagree. I followed the standard how slack also does that. The duplication would confuse the user.

> * I was able to add users to a channel without being the channel admin. (Maybe because I am an Artemis admin?) The separation is unclear, maybe rename to "Channel Owners" or similar.

That is the case, yes. Instructors have channel administration rights but are not necessarily channel admins. A course normally only has one instructor so I do not think its that confusing for him or her? And as always in Artemis. The admin can do everything.


> * Message search: Should work by pressing enter and also highlight the text.

I agree but the whole message part was done by @canberkanar . I can maybe work on this on a follow up. But I did not spend much time to look at that code yet.

> * When pressing enter in the message field, it should be disabled until the request was successful (can still enter text while sending).

I agree but the whole message part was done by @canberkanar . I can maybe work on this on a follow up. But I did not spend much time to look at that code yet.


> * Showing the ID under details is unnecessary for users.

I agree but it is important for support purposes. Thats why also slack shows it in the Info tab.


> * Wording: Sometimes the UI uses "channel" and other times "conversation".

Can you give an example? OneToOneChat, GroupChat, Channel are all instantiations of a conversation.

tobias-lippert reviewed (commented):
Great work.  Works as expected. Only discussed minor improvements in the testing session.

danielbode reviewed (commented):
Really nice PR! Tested on ts1 during a testing session and didn't notice any bugs. We discussed some minor usability improvements:
- add setting to choose whether you send or new line on enter (Might be useful for larger announcements so you don't accidentally send the message)
- add possibility to remove DMs (Removing is stronger than hiding. After removing, chats don't collapse in the hidden view, but are gone until you message that user again).

pellumbbaboci reviewed (commented):
Test on ts1 during testing session works as expected, and I did not find any bugs.

stefanwaldhauser commented:
As requested in the testing session, I made the left sidebar scrollable if the user has a lot of conversations in commit 0ddcd7d817927ee3012e724e38b150a021feffab
![image](https://user-images.githubusercontent.com/29718932/207268822-d18645fa-c7f9-4df5-8ddf-418f3b9267a9.png)

stefanwaldhauser commented:
> Really nice PR! Tested on ts1 during a testing session and didn't notice any bugs. We discussed some minor usability improvements:
> 
> * add setting to choose whether you send or new line on enter (Might be useful for larger announcements so you don't accidentally send the message)
> * add possibility to remove DMs (Removing is stronger than hiding. After removing, chats don't collapse in the hidden view, but are gone until you message that user again).

Discussed this with @krusche and will be done in a follow up

danielbode reviewed (approved):
Tested on ts2 again after your changes. Looks good!

I have one more remark, that you can maybe address in a follow-up. If you create a group chat or add users to a channel, it doesn't feel natural to get an error message, that you have to select at least one user, when only changing filter options (see attached screencast). If we want to keep this feedback to the user, we can maybe place it less prominent behind an icon with a popover like we do for exercise creation.

https://user-images.githubusercontent.com/44734881/207289328-e1352229-fdb1-47e0-80de-05d3d6222308.mov

stefanwaldhauser commented:
> Tested on ts1. Scrollbar works as expected. Found one issue: If one (in this case a student, didn't test it with other roles) wants to create a group chat and adds more than 9 members to the participants list and removes one to not be above the limit of 10, the creation fails with a bad request. ![image](https://user-images.githubusercontent.com/84102468/207302796-99af384d-4281-4c99-b8c4-b0a40efcd998.png)

Fixed bug in cb55f929969656b62b27c00051d3c3a76f75cc9f

tobias-lippert reviewed (approved):
Bug is fixed. Great work!

pellumbbaboci reviewed (approved):
Re-approved, everything works as expected.

evgeny1984 reviewed (changes requested):
When I add a new user from direct messages to the favorite list, he or she disappears from the direct message list. I think, these lists are independent and the user should be in both.

danielbode reviewed (approved):
Reapprove after changes. Still looks good!

stefanwaldhauser commented:
> When I add a new user from direct messages to the favorite list, he or she disappears from the direct message list. I think, these lists are independent and the user should be in both.

I followed what other chat apps like Slack do. They also move the whole conversation instead of duplicating it. Is that a blocker for you?

krusche commented:
Let's get at least 6 approvals for this huge PR

evgeny1984 reviewed (changes requested):
Tested on ts2, unfortunately found an issue with internal server error.
![image](https://user-images.githubusercontent.com/10652693/208123532-e2e41956-b811-42d9-90ae-7f1163e50c23.png)
Steps to reproduce:
Write message to another user → reply in thread → react with emoji on the answer

ge65cer commented on src/main/java/de/tum/in/www1/artemis/domain/metis/ConversationParticipant.java:
```diff
@@ -29,6 +30,18 @@ public class ConversationParticipant extends DomainObject {
+     * Currently only used for {@link de.tum.in.www1.artemis.domain.metis.conversation.Channel}
+     */
+    @Column(name = "is_admin")
+    private Boolean isAdmin;
```
See my previous comment, the term "admin" in channels is ambiguous with our definition of a "user admin".

stefanwaldhauser replied:
> Tested on ts2; unfortunately, I found an issue with an internal server error. ![image](https://user-images.githubusercontent.com/10652693/208123532-e2e41956-b811-42d9-90ae-7f1163e50c23.png) Steps to reproduce: Write message to another user → reply in thread → react with emoji on the answer

I could not reproduce this error, but an existing error causes an internal server error when you too quickly add too many reactions or multiple users do it simultaneously. It was not introduced in this PR. Is it possible that it is this bug? Or does it happen with every reaction, also meaning if you add them slowly? 



https://user-images.githubusercontent.com/29718932/208307169-6e14ec31-1e79-426d-b7bb-971d86a858c1.mp4

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/dto/UserPublicInfoDTO.java:
```diff
@@ -0,0 +1,165 @@
+ * A DTO representing a user with the minimal information allowed to be seen by other users in a course
+ */
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public class UserPublicInfoDTO {
```
I think we switched to records?
https://www.baeldung.com/java-record-keyword

stefanwaldhauser replied:
Records are just syntactic sugar. I see no reasons to change this here.

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/ConversationMessagingService.java:
```diff
@@ -182,7 +191,13 @@ private Conversation mayUpdateOrDeleteMessageElseThrow(Post existingMessagePost,
-            return conversationService.getConversationById(existingMessagePost.getConversation().getId());
+            var conversation = conversationService.getConversationById(existingMessagePost.getConversation().getId());
+            if (conversation instanceof Channel channel) {
+                if (channel.getIsArchived()) {
```
Codacy warning

stefanwaldhauser replied:
Fixed in 0dd03f26a7f0b4414b762f40b6e9bbcca87798ec

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java:
```diff
@@ -0,0 +1,224 @@
+     * @param channelId the id of the channel
+     * @return the channel with the given id
+     */
+    public Channel getChannelOrThrow(Long channelId) {
```
Should be in Repository?

stefanwaldhauser replied:
Done in 60be6e3895a0e582d0d73f1312ef2a66fb4dbc51

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java:
```diff
@@ -0,0 +1,224 @@
+     * @param courseId the id of the course
+     * @return a list of channels for the given course
+     */
+    public List<Channel> getChannels(Long courseId) {
```
Call the repository method directly?

stefanwaldhauser replied:
I do not like to break the Resource -> Service -> Repository layer that often tbh. I think its not "bad" to have it in the service. So I do not need to add a new dependency to the resource

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java:
```diff
@@ -0,0 +1,224 @@
+     * @param courseId the id of the course
+     * @param channel  the channel to check
+     */
+    public void channelIsValidOrThrow(Long courseId, Channel channel) {
```
The parameter annotation `@Valid` could simply work as well?

stefanwaldhauser replied:
Good idea! Done in: f4b71e3d764609ac40b8fdfd1ad3f67f1eebcdf5

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java:
```diff
@@ -0,0 +1,224 @@
+     *
+     * @param channel the channel to delete
+     */
+    public void deleteChannel(Channel channel) {
```
Call the repository method directly?

stefanwaldhauser replied:
I do not like to break the Resource -> Service -> Repository layer that often tbh. I think its not "bad" to have it in the service. So I do not need to add a new dependency to the resource

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/GroupChatService.java:
```diff
@@ -0,0 +1,120 @@
+     * @return the group chat
+     */
+    public GroupChat getGroupChatOrThrow(Long groupChatId) {
+        return groupChatRepository.findById(groupChatId)
```
Same as above.

stefanwaldhauser replied:
I do not like to break the Resource -> Service -> Repository layer that often tbh. I think its not "bad" to have it in the service. So I do not need to add a new dependency to the resource

stefanwaldhauser replied:
Done in a213247889b860e291776ad1627056bcd7eb77c0

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/GroupChatService.java:
```diff
@@ -0,0 +1,120 @@
+     *
+     * @param groupChat the group chat to check
+     */
+    public void groupChatIsValidOrThrow(GroupChat groupChat) {
```
Same as above.

stefanwaldhauser replied:
Done in 37585c59926919519091389f87bcd31d8c84e40c

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/GroupChatService.java:
```diff
@@ -0,0 +1,120 @@
+     * @param groupChatDTO the DTO containing the new group chat data
+     * @return the updated group chat
+     */
+    public GroupChat updateGroupChat(Long groupChatId, GroupChatDTO groupChatDTO) {
```
This logic is only related to storing an updated group chat? Could be directly done in the Resource class.

stefanwaldhauser replied:
I would like to keep the update logic in the service as with the other types.

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/OneToOneChatService.java:
```diff
@@ -0,0 +1,83 @@
+     * @return the OneToOneChat if it exists, otherwise empty optional
+     */
+    public Optional<OneToOneChat> findOneToOneChatWithSameMembers(Long courseId, Long userAId, Long userBId) {
+        return oneToOneChatRepository.findBetweenUsersWithParticipantsAndUserGroups(courseId, userAId, userBId);
```
Same as above.

stefanwaldhauser replied:
Done in 8eae2ed6fb3b63c7c7595178f24dfb387a25e4b4

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/auth/GroupChatAuthorizationService.java:
```diff
@@ -0,0 +1,95 @@
+        var userToCheck = getUserIfNecessary(user);
+        var createdGroupChats = groupChatRepository.countByCreatorIdAndCourseId(userToCheck.getId(), course.getId());
+        if (createdGroupChats >= MAX_GROUP_CHATS_PER_USER_PER_COURSE) {
+            throw new IllegalArgumentException("You can only create " + MAX_GROUP_CHATS_PER_USER_PER_COURSE + "group chats per course");
```
```suggestion
            throw new IllegalArgumentException("You can only create " + MAX_GROUP_CHATS_PER_USER_PER_COURSE + " group chats per course");
```

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/auth/OneToOneChatAuthorizationService.java:
```diff
@@ -0,0 +1,44 @@
+        var userToCheck = getUserIfNecessary(user);
+        var createdOneToOneChats = oneToOneChatRepository.countByCreatorIdAndCourseId(userToCheck.getId(), course.getId());
+        if (createdOneToOneChats >= MAX_ONE_TO_ONE_CHATS_PER_USER_PER_COURSE) {
+            throw new AccessForbiddenException("You can only create " + MAX_ONE_TO_ONE_CHATS_PER_USER_PER_COURSE + "one to one chats per course");
```
```suggestion
            throw new AccessForbiddenException("You can only create " + MAX_ONE_TO_ONE_CHATS_PER_USER_PER_COURSE + " one to one chats per course");
```

ge65cer commented on src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:
```diff
@@ -828,7 +878,7 @@ public ResponseEntity<String> getCourseTitle(@PathVariable Long courseId) {
      * Post /courses/:courseId/students/:studentLogin : Add the given user to the students of the course so that the student can access the course
      *
-     * @param courseId     the id of the course
+     * @param courseId     the id of the course f
```
```suggestion
     * @param courseId     the id of the course
```

ge65cer commented on src/main/java/de/tum/in/www1/artemis/web/rest/metis/conversation/dtos/ChannelDTO.java:
```diff
@@ -0,0 +1,115 @@
+
+    // property not taken from entity
+    /**
+     * Am course instructor has channel admin rights but is not necessarily a member or admin of the channel
```
```suggestion
     * A course instructor has channel admin rights but is not necessarily a member or admin of the channel
```

ge65cer commented on src/main/java/de/tum/in/www1/artemis/web/rest/metis/conversation/dtos/ConversationDTO.java:
```diff
@@ -0,0 +1,163 @@
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+@JsonSubTypes({ @JsonSubTypes.Type(value = OneToOneChatDTO.class, name = "oneToOneChat"), @JsonSubTypes.Type(value = GroupChatDTO.class, name = "groupChat"),
+        @JsonSubTypes.Type(value = ChannelDTO.class, name = "channel"), })
+public class ConversationDTO {
```
Could also be a record?

stefanwaldhauser replied:
Sadly not. Records already extend [java.lang.Record](https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/lang/Record.html). As Java does not allow multiple inheritance, records cannot extend any other class.

stefanwaldhauser replied:
I could do it with this parent class but would be weird IMO if the parent class is a record but the subclasses not.

ge65cer commented on src/main/webapp/app/overview/course-conversations/dialogs/channels-create-dialog/channel-form/channel-form.component.ts:
```diff
@@ -0,0 +1,97 @@
+    ngOnDestroy() {
+        this.ngUnsubscribe.next();
+        this.ngUnsubscribe.complete();
+    }
```
```suggestion
    }
    
```

ge65cer commented on src/main/webapp/app/overview/course-conversations/dialogs/channels-create-dialog/channels-create-dialog.component.ts:
```diff
@@ -0,0 +1,41 @@
+export class ChannelsCreateDialogComponent extends AbstractDialogComponent {
+    @Input()
+    course: Course;
+    initialize() {
```
```suggestion

    initialize() {
```

ge65cer commented:
I did a first code review, but we desperately need more eyes due to the many added lines.

stefanwaldhauser commented:
> Tested on ts2, unfortunately found an issue with internal server error. ![image](https://user-images.githubusercontent.com/10652693/208123532-e2e41956-b811-42d9-90ae-7f1163e50c23.png) Steps to reproduce: Write message to another user → reply in thread → react with emoji on the answer

I could not reproduce this error, but an existing error causes an internal server error when you too quickly add too many reactions or multiple users do it simultaneously. It was not introduced in this PR. Is it possible that it is this bug? Or does it happen with every reaction, also meaning if you add them slowly?


https://user-images.githubusercontent.com/29718932/208307209-3bf23a3f-15d9-4b46-83fe-5da03d5cd50c.mp4

danielbode reviewed (approved):
Tested again on ts1 in a testing session. The feature works well, I didn't notice any bugs.

Yunqing-Wang reviewed (approved):
works perfectly, tested in the test session

rriyaldhi reviewed (dismissed):
Tested on TS1. Functionality works well!

YeetTheFirst21 reviewed (approved):
Looks good, joined the test session. Little visibility changes like color and border changes for messaging part would be nice in order to help visually declare where the messaging window begins (was gone over in the test session).

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java:
```diff
@@ -0,0 +1,185 @@
+        var conversationParticipantOfRequestingUser = new ConversationParticipant();
+        conversationParticipantOfRequestingUser.setUser(user);
+        conversationParticipantOfRequestingUser.setConversation(savedChannel);
+        // creator is of moderator. Special case, because creator is the only moderator that can not be revoked the role
```
```suggestion
        // Creator is a moderator. Special case, because creator is the only moderator that can not be revoked the role
```

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java:
```diff
@@ -0,0 +1,393 @@
+        conversations.addAll(oneToOneChatsOfUser);
+        conversations.addAll(channelsOfUser);
+        conversations.addAll(groupChatsOfUser);
+        return conversations.stream().map(conversation -> conversationDTOService.convertToDTO(conversation, requestingUser)).collect(Collectors.toList());
```
[Docs](https://docs.artemis.cit.tum.de/dev/guidelines/server/?highlight=tolist#general-best-practices)
```suggestion
        return conversations.stream().map(conversation -> conversationDTOService.convertToDTO(conversation, requestingUser)).toList();
```

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java:
```diff
@@ -0,0 +1,393 @@
+     */
+    public ZonedDateTime auditConversationReadTimeOfUser(Conversation conversation, User user) {
+        // update the last time user has read the conversation
+        ConversationParticipant readingParticipant = conversationParticipantRepository.findConversationParticipantByConversationIdAndUserId(conversation.getId(), user.getId())
```
The method for `orElseThrow` should be in the Repository.

stefanwaldhauser replied:
Done in 1db0a029fc229bb2463a3fae195117c3a1f3aa81

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java:
```diff
@@ -0,0 +1,393 @@
+     * @param isFavorite     the new favorite status
+     */
+    public void switchFavoriteStatus(Long conversationId, User requestingUser, Boolean isFavorite) {
+        var participation = conversationParticipantRepository.findConversationParticipantByConversationIdAndUserId(conversationId, requestingUser.getId())
```
The method for `orElseThrow` should be in the Repository.

stefanwaldhauser replied:
Done in 1db0a029fc229bb2463a3fae195117c3a1f3aa81

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java:
```diff
@@ -0,0 +1,393 @@
+     * @param hiddenStatus   the new hidden status
+     */
+    public void switchHiddenStatus(Long conversationId, User requestingUser, Boolean hiddenStatus) {
+        var participation = conversationParticipantRepository.findConversationParticipantByConversationIdAndUserId(conversationId, requestingUser.getId())
```
The method for `orElseThrow` should be in the Repository.

stefanwaldhauser replied:
Done in 1db0a029fc229bb2463a3fae195117c3a1f3aa81

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/auth/ChannelAuthorizationService.java:
```diff
@@ -0,0 +1,265 @@
+        var isArchivedChannel = !Objects.isNull(channel.getIsArchived()) && channel.getIsArchived();
+        var userToCheck = getUserIfNecessary(user);
+        var isChannelMember = isMember(channel.getId(), userToCheck.getId());
+        if (isArchivedChannel) {
```
Maybe move this two lines above, so the method skips early instead of loading the user?

stefanwaldhauser replied:
Good idea! done in b65b58c89ec5b14e3980c47be531797a225b85ba

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/auth/ChannelAuthorizationService.java:
```diff
@@ -0,0 +1,265 @@
+        var isArchivedChannel = !Objects.isNull(channel.getIsArchived()) && channel.getIsArchived();
+        var userToCheck = getUserIfNecessary(user);
+
+        if (isArchivedChannel) {
```
Same as above.

stefanwaldhauser replied:
Done in 7321b8ac348e12dee95c35679b2474892782d703

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/auth/ConversationAuthorizationService.java:
```diff
@@ -0,0 +1,43 @@
+        this.authorizationCheckService = authorizationCheckService;
+    }
+
+    protected User getUserIfNecessary(User user) {
```
Needs javadoc.

stefanwaldhauser replied:
Added javadoc in this commit: 4acaf46ab01ba755be935993dd4e617f1baf1129

ge65cer commented on src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:
```diff
@@ -748,6 +751,53 @@ public ResponseEntity<List<UserDTO>> searchStudentsInCourse(@PathVariable Long c
+    @PreAuthorize("hasRole('USER')")
+    public ResponseEntity<List<UserPublicInfoDTO>> searchUsersInCourse(@PathVariable Long courseId, @RequestParam("loginOrName") String loginOrName,
+            @RequestParam("roles") List<String> roles) {
+        // ToDo: Discuss what information should be hidden as this is a public endpoint
```
Open todo.

stefanwaldhauser replied:
> Looks good, joined the test session. Little visibility changes like color and border changes for messaging part would be nice in order to help visually declare where the messaging window begins (was gone over in the test session).

I adapted the color and added a border to the messaging window. As requested in the testing session I also added the ability of channel moderators to edit and delete messages in a channel.

stefanwaldhauser replied:
Removed the todo in 7774f6c5d3d4504aa56ec1626100926ac97057f7

Already discussed. We are sending the minimal information which just includes login and the name. With less we can not differentiate between users with the same name in the chat.

ge65cer reviewed (changes requested):
Reviewed server code, found only a few more small issues. 🙂

stefanwaldhauser commented:
> Looks good, joined the test session. Little visibility changes like color and border changes for messaging part would be nice in order to help visually declare where the messaging window begins (was gone over in the test session).

I adapted the color and added a border to the messaging window. As requested in the testing session I also added the ability of channel moderators to edit and delete messages in a channel.

stefanwaldhauser commented:
> Reviewed server code, found only a few more small issues. 🙂

Thanks for the review :) Fixed all the issues

evgeny1984 reviewed (changes requested):
Unfortunately, error in the console with object null reference exception. Steps to reproduce: Go to the course → messages → search for not existing user and then close the dialog.
![image](https://user-images.githubusercontent.com/10652693/208661417-7cd6c6d8-6ecf-442b-8d0e-6d32b848f381.png)

stefanwaldhauser commented:
> Unfortunately, error in the console with object null reference exception. Steps to reproduce: Go to the course → messages → search for not existing user and then close the dialog. ![image](https://user-images.githubusercontent.com/10652693/208661417-7cd6c6d8-6ecf-442b-8d0e-6d32b848f381.png)

Thats not really a bug per so but expected behaviour. 

The error you are seeing in the console is the one from a browser indicating that a promise was rejected and there was no handler for it. So in this sense it works as expected - you need to handle a promise if you don't want to see the default error for an unhandled promise rejection.

https://github.com/ng-bootstrap/ng-bootstrap/blob/b8230e5cb39b262c61c118e50e1cc5e33a334f96/src/modal/modal-ref.ts#L42-L45

NgBootstrap sadly rejects the result promise if the modal is dismissed.

Anyway to not spam sentry, I added a catch in commit ad58cb24e83d325cffb3d87819c64f122fad7a25 so the "error" should not show up in the console any more :)

stefanwaldhauser commented:
Failing server tests are unrelated to this PR. I just checked the logs.

ge65cer commented on src/test/javascript/spec/helpers/mocks/directive/ngbPaginationMocks.module.ts:
```diff
@@ -1,23 +1,35 @@
 
+// eslint-disable-next-line @angular-eslint/directive-selector
 @Directive({ selector: 'ng-template[ngbPaginationLast]' })
+// eslint-disable-next-line @angular-eslint/directive-class-suffix
```
Please don't disable the linter here, but rather fix the underlying error (adding the 'Directive' suffix).

stefanwaldhauser replied:
Done in f14078c614f7f4a4e43952c7bb56f674b71e0bf1

ge65cer commented on src/main/java/de/tum/in/www1/artemis/repository/metis/conversation/GroupChatRepository.java:
```diff
@@ -0,0 +1,37 @@
+
+    @EntityGraph(type = LOAD, attributePaths = { "conversationParticipants.user.groups" })
+    @Query("""
+                     SELECT DISTINCT groupChat
```
```suggestion
                 SELECT DISTINCT groupChat
```

ge65cer commented on src/main/java/de/tum/in/www1/artemis/repository/metis/conversation/OneToOneChatRepository.java:
```diff
@@ -0,0 +1,57 @@
+
+    @EntityGraph(type = LOAD, attributePaths = { "conversationParticipants.user.groups" })
+    @Query("""
+            SELECT DISTINCT  o FROM OneToOneChat o
```
```suggestion
            SELECT DISTINCT o FROM OneToOneChat o
```

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/ConversationMessagingService.java:
```diff
@@ -0,0 +1,232 @@
+            }
+        }
+        conversationParticipantRepository.incrementUnreadMessagesCountOfParticipants(conversation.getId(), author.getId());
+        // ToDo: Optimization Idea: Maybe we can save this websocket call and instead get the last message date from the conversation object in the post somehow?
```
Open todo.

stefanwaldhauser replied:
Will be done in a follow up. I sometimes write todos for optimization ideas

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/ConversationMessagingService.java:
```diff
@@ -0,0 +1,232 @@
+        Post existingMessage = conversationMessageRepository.findMessagePostByIdElseThrow(postId);
+        Conversation conversation = mayUpdateOrDeleteMessageElseThrow(existingMessage, user);
+
+        // ToDo: find a cleaner way to do this instead of making the string here in the server
```
Open todo. I agree, we should not create this string on the server but instead let the client handle the translation.

stefanwaldhauser replied:
Will be done in a follow up.

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java:
```diff
@@ -0,0 +1,189 @@
+    public Channel updateChannel(Long channelId, Long courseId, ChannelDTO channelDTO) {
+        var channel = channelRepository.findByIdElseThrow(channelId);
+        if (channelDTO.getName() != null && !channelDTO.getName().equals(channel.getName())) {
+            channel.setName(channelDTO.getName().trim().isBlank() ? "" : channelDTO.getName().trim());
```
Hmm... why the distinction and not simply `channel.setName(channelDTO.getName().trim())`?

stefanwaldhauser replied:
True! Changed it in commit: b458ed39de275ed1eb7057c8a9e68610a94f6d31

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/OneToOneChatService.java:
```diff
@@ -0,0 +1,72 @@
+        participant.setIsHidden(false);
+        participant.setIsFavorite(false);
+        // set the last reading time of a participant in the past when creating conversation for the first time!
+        participant.setLastRead(ZonedDateTime.now().minusYears(2));
```
Why exactly minus two years?

stefanwaldhauser replied:
I am not exactly sure. This was done by @pellumbbaboci and I just migrated it into this Pr.

stefanwaldhauser replied:
I think so that the existing messages show up as unread?

ge65cer replied:
The method is only called when a conversation is created, so there will not be any existing messages. Right?

pellumbbaboci replied:
Yes @ge65cer, its only called once. And there is no reason for minus 2 years i just need the date to be in the past and not null.

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/auth/ConversationAuthorizationService.java:
```diff
@@ -0,0 +1,50 @@
+     * @return the {@link User} object, potentially after loading the authorities and groups
+     * @throws IllegalArgumentException if the user parameter is null
+     */
+    protected User getUserIfNecessary(User user) {
```
Shouldn't this maybe be in the User/Auth Service? I understand it is a helper method, but it does not perform anything specific to the Conversation Service.

stefanwaldhauser replied:
The corresponding method in the auth service is private and is slightly different as it assumes if user is null then the currently requesting user is meant. I would rather not introduce another public method for my case here.

ge65cer commented on src/main/java/de/tum/in/www1/artemis/web/rest/metis/conversation/ConversationResource.java:
```diff
@@ -0,0 +1,167 @@
+    @GetMapping("/{courseId}/conversations")
+    @PreAuthorize("hasRole('USER')")
+    public ResponseEntity<List<ConversationDTO>> getConversationsOfUser(@PathVariable Long courseId) {
+        var requestingUser = this.userRepository.getUserWithGroupsAndAuthorities();
```
We normally do not add `this` when the method signature does not have a conflicting parameter name (applies to the whole class and everywhere else).
```suggestion
        var requestingUser = userRepository.getUserWithGroupsAndAuthorities();
```

ge65cer commented on src/main/java/de/tum/in/www1/artemis/web/rest/metis/conversation/GroupChatResource.java:
```diff
@@ -0,0 +1,181 @@
+        groupChatAuthorizationService.isAllowedToRemoveUsersFromGroupChat(groupChatFromDatabase, requestingUser);
+        var usersToDeRegister = conversationService.findUsersInDatabase(userLogins);
+        conversationService.deregisterUsersFromAConversation(course, usersToDeRegister, groupChatFromDatabase);
+        // ToDo: Discuss if we should delete the group chat if it has no participants left, but maybe we want to keep it for data analysis purposes
```
Open todo.

stefanwaldhauser replied:
Will be done in a follow up.

ge65cer commented on src/main/java/de/tum/in/www1/artemis/web/rest/metis/conversation/dtos/ChannelDTO.java:
```diff
@@ -0,0 +1,115 @@
+
+    private Boolean isArchived;
+
+    // property not taken from entity
```
What does that mean?

stefanwaldhauser replied:
Tthat it is not taken from the entity but calculated / based on other entities

ge65cer commented on src/main/resources/config/liquibase/changelog/20221225120000_changelog.xml:
```diff
@@ -0,0 +1,49 @@
+            <column name="is_moderator" type="boolean" defaultValueBoolean="false">
+                <constraints nullable="true" />
+            </column>
+            <column name="is_hidden" type="boolean" defaultValueBoolean="false">
```
I don't quite understand this here... You are creating a ConversationParticipant for every user in a conversation.
How do you track the hidden/favorite flag per user then? Wouldn't that require an n-to-n relation?

stefanwaldhauser replied:
```
                     Contains user specific settings for the conversation
                            │
                            │
                            │
                            │
                            │
     1         n            ▼            n      1
User   ────────  ConversationParticipant  ──────  Conversation
```

stefanwaldhauser replied:
I do not like the name participant to be fair. I think "Subscription" would be better, so basically you can use this entity to store user specific settings per conversation.

stefanwaldhauser replied:
Added uniqueness constraint in eab79f17261a0b27078cbc9ac54d22f552dcd1ce to enforce this

ge65cer replied:
Okay, I think my misunderstanding was that the "hidden" and "favorite" flags were for the user, when actually they belong to the conversation.

ge65cer replied:
I think you already have such a constraint? https://github.com/ls1intum/Artemis/pull/5822/files/7915be8619ff2bbfa5c9eeab196c0713ee625901#diff-f9d1bc6e238168a0ff3a44a1a08d541ecf1c0adccda8a490a1b5853b22e17635R42

stefanwaldhauser replied:
Yeah removed it again in the latest commit :)

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationDTOService.java:
```diff
@@ -0,0 +1,208 @@
+        var participantOptional = conversationParticipantRepository.findConversationParticipantByConversationIdAndUserId(channel.getId(), requestingUser.getId());
+        channelDTO.setIsMember(participantOptional.isPresent());
+        participantOptional.ifPresent(conversationParticipant -> channelDTO.setLastReadDate(conversationParticipant.getLastRead()));
+        participantOptional.ifPresent(conversationParticipant -> channelDTO.setUnreadMessagesCount(conversationParticipant.getUnreadMessagesCount()));
```
You can optimize this by merging the two lines.

stefanwaldhauser replied:
Done in 11008eb152db6fbddc6487737ac0b019b0b742b7

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationDTOService.java:
```diff
@@ -0,0 +1,208 @@
+        participantOfRequestingUser.ifPresent(conversationParticipant -> groupChatDTO.setLastReadDate(conversationParticipant.getLastRead()));
+        participantOfRequestingUser.ifPresent(conversationParticipant -> groupChatDTO.setUnreadMessagesCount(conversationParticipant.getUnreadMessagesCount()));
+        participantOfRequestingUser.ifPresent(conversationParticipant -> groupChatDTO.setIsFavorite(conversationParticipant.getIsFavorite()));
+        participantOfRequestingUser.ifPresent(conversationParticipant -> groupChatDTO.setIsHidden(conversationParticipant.getIsHidden()));
```
Same as above.

stefanwaldhauser replied:
Done in 11008eb152db6fbddc6487737ac0b019b0b742b7

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationDTOService.java:
```diff
@@ -0,0 +1,208 @@
+        participantOfRequestingUser.ifPresent(conversationParticipant -> oneToOneChatDTO.setLastReadDate(conversationParticipant.getLastRead()));
+        participantOfRequestingUser.ifPresent(conversationParticipant -> oneToOneChatDTO.setUnreadMessagesCount(conversationParticipant.getUnreadMessagesCount()));
+        participantOfRequestingUser.ifPresent(conversationParticipant -> oneToOneChatDTO.setIsFavorite(conversationParticipant.getIsFavorite()));
+        participantOfRequestingUser.ifPresent(conversationParticipant -> oneToOneChatDTO.setIsHidden(conversationParticipant.getIsHidden()));
```
You can optimize this by merging the above lines.

ge65cer commented on src/main/java/de/tum/in/www1/artemis/repository/metis/conversation/OneToOneChatRepository.java:
```diff
@@ -0,0 +1,57 @@
+            AND
+            EXISTS (SELECT p FROM ConversationParticipant p WHERE p.user.id = :#{#userIdA} AND p.conversation = o)
+            AND
+            EXISTS (SELECT p FROM ConversationParticipant p WHERE p.user.id = :#{#userIdB} AND p.conversation = o)
```
Are these sub-queries really required? A join might be better performance-wise.

stefanwaldhauser replied:
Removed subquery in 3846cca54aad0a5c67d37f7915667103fa030fe0

ge65cer reviewed (changes requested):
A final code review of the changes to the server component. 🙂

simon-hng commented on src/main/java/de/tum/in/www1/artemis/service/dto/UserPublicInfoDTO.java:
```diff
@@ -0,0 +1,165 @@
+
+    private String lastName;
+
+    private Boolean isInstructor;
```
Change to enum instead of using multiple boolean flags

stefanwaldhauser replied:
Will be done in a follow up

TheZoker reviewed (approved):
Tested on TS1 during testing session (no code review yet). Works, great addition! 🎉

akesfeden reviewed (approved):
Tested on TS1 during testing session. Sending/receiving messages via direct/channel/group chats, seeing unread message count, editing a message in a channel by the instructor works as expected.

pellumbbaboci reviewed (approved):
Tested during testing session on TS1, everything works fine :+1:

simon-hng reviewed (approved):
Tested on TS1. One small issue with Brave browser, however it is not officially supported

stefanwaldhauser commented:
> Tested on TS1. One small issue with Brave browser, however it is not officially supported

Maybe this helps: https://community.brave.com/t/brave-browser-shields-blocking-my-websocket/395377 ?

ge65cer reviewed (changes requested):
There are open discussions. The client code has not been reviewed by anyone, yet.
